### PR TITLE
SDL Compliance Completion

### LINKED ISSUE #6918: SDL compliance
There are a bunch of warnings that are being suppressed (esp in RN core and folly).
We should identify them, and fix them.

### PR DESCRIPTION
**Pull request by AgneLukoseviciute**
Adds two forked folly files in order to re-enable C4244 and C4267 warning suppressions. I opened a [PR in folly](https://github.com/facebook/folly/pull/1565) to address this from there, and will add an issue to track updating folly and removing these files. 
Closes #6918. 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7659)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj [hunk 1]
```diff
@@ -101,7 +101,7 @@
       <PrecompiledHeaderFile>pch.h</PrecompiledHeaderFile>
       <PrecompiledHeaderOutputFile>$(IntDir)pch.pch</PrecompiledHeaderOutputFile>
       <ForcedIncludeFiles>pch.h</ForcedIncludeFiles>
-      <WarningLevel>Level4</WarningLevel>
+      <WarningLevel>Level3</WarningLevel>
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj:104]: Why is this needed? Seems like this will disable a bunch of warnings

File: vnext/Folly/TEMP_UntilFollyUpdate/lang/ToAscii.h [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+using to_ascii_alphabet_upper = to_ascii_alphabet<true>;
+
+namespace detail {
+
+FOLLY_ERASE int to_ascii_port_clzll(uint64_t v) {
+#if _MSC_VER
+#if FOLLY_X64
+  return static_cast<int>(__lzcnt64(v));
+#else
+  __assume(0);
```
Comment (asklar) [inline vnext/Folly/TEMP_UntilFollyUpdate/lang/ToAscii.h:64]: Do x86 builds use FOLLY_X64?

**Review by asklar — CHANGES_REQUESTED** [review]
Left some questions. 

Meta point (this one shouldn't block this PR), it might be better to use patches instead of forked files to be able to PR the individual forked change rather than an entire file that has maybe one line different. Thoughts?

File: vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj (hunk 1 above)
Comment (AgneLukoseviciute) [reply vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj:104]: Not needed, I'll change that back.

File: vnext/Folly/TEMP_UntilFollyUpdate/lang/ToAscii.h (hunk 2 above)
Comment (AgneLukoseviciute) [reply vnext/Folly/TEMP_UntilFollyUpdate/lang/ToAscii.h:64]: From working on this it appears to be, but I'm currently trying to figure out how it works exactly for the fix in folly b/c the return type of this method changed to auto but moving the cast to the caller on line 227 only works in x64. x86 is complaining about trying to convert void to int.

**Comment by AgneLukoseviciute** [comment]
> Left some questions.
> 
> Meta point (this one shouldn't block this PR), it might be better to use patches instead of forked files to be able to PR the individual forked change rather than an entire file that has maybe one line different. Thoughts?

@asklar That sounds like a better approach, especially with the minimal changes. I can update this to use patches instead.

**Comment by asklar** [comment]
[talked offline, we still need to figure out how x86 should/will work]

**Comment by AgneLukoseviciute** [comment]
@asklar, I updated to_ascii_port_clzll() method to call builtin_clzll() on x86. builtin_clzll() will then call __builtin_clz() which is supported only in VS 16.3-16.8 but code added [in folly](https://github.com/facebook/folly/pull/1450/files) handles the logic for calling polyfill that uses _BitScanReverse when __builtin_clz is not available

With the forked files it's hard to see the diffs; for reference the only changes I've made to these files from what we were using before are in ToAscii.h lines 59-63(adding check for x86), 64/96(updates to match changes upstream - return type changed to auto from int and added return statement in line 69), 84(added cast), & 233(added cast). in Dynamic.cpp the only change is an added cast on line 300.

**Review by asklar — APPROVED** [review]
